### PR TITLE
Don't overwrite torch::nn::Sequential module names when cloning

### LINKED ISSUE #71069: C++ torch::nn::Sequential clone() method overwrites child module names
### 🐛 Describe the bug

The clone method of torch::nn::Sequential fails to pick up the names of child modules, e.g.

``` 
torch::nn::Sequential q({ {"A", torch::nn::Linear(1,2)},
                           {"B", torch::nn::ReLU()} });
 std::cerr << *q << "\n";

 auto c=q->clone();
 std::cerr << *c << "\n";
```
The above code results in:
```
torch::nn::Sequential(
  (A): torch::nn::Linear(in_features=1, out_features=2, bias=true)
  (B): torch::nn::ReLU()
)

torch::nn::Sequential(
  (0): torch::nn::Linear(in_features=1, out_features=2, bias=true)
  (1): torch::nn::ReLU()
)
```

the code in **_include/torch/nn/modules/container/sequential.h_** uses **_push_back_** without name:
```
  std::shared_ptr<Module> clone(
      const optional<Device>& device = nullopt) const override {
    auto clone = std::make_shared<SequentialImpl>();
    for (const auto& module : modules_) {
      clone->push_back(module.clone(device));
    }
    return clone;
  }
```

I think the clone loop needs to use the name of the child modules in the **_push_back()_** call,
something like:
```
    for (const auto& m : named_children())
      clone->push_back(m.key(), AnyModule(m.value().clone(device)));
```

### Versions

I think this bug has been part of torch::nn::Sequential from the beginning 

cc @jbschlosser @albanD @mruberry @walterddr @kshitij12345

### PR DESCRIPTION
**Pull request by kvathupo**
__What?__
Fixes #71069

Further changes:
1. Changed the parameters of the private `torch::nn::Sequential::push_back()` method used to unpack the parameters from the variadic constructor
2. Added documentation to said constructor

__Why?__
1. I initially changed the private `push_back()` method to solve what turned out to be a non-issue. Afterwards, I found these changes made it easier to understand. Also, now we don't need to rely on any public `push_back()` methods in unpacking: originally, the base case of 1 parameter in the variadic template was covered by the public `push_back(M&& module)`.

__Further Development?__
Possible changes with C++17 (lmk if there's a wish list for that):
* Replace `c10::optional` with `std::optional`
* Add fine-grained type-checking with type traits in `class Sequential`; will save following the dependency chain

__P.S.__
I think you guys squash commits; if not, I can smash them.

### CODE REVIEW

**Comment by kvathupo** [comment]
@malfet ping

**Review by malfet — COMMENTED** [review]
Hmm, what would happen if one tries to clone container with numerical object names

**Comment by kvathupo** [comment]
> Hmm, what would happen if one tries to clone container with numerical object names

If I'm understanding correctly, that case is covered in the test: it mixes named and unnamed modules (modules of the sequence take the respective index value in the unnamed case)

File: torch/csrc/api/include/torch/nn/modules/container/sequential.h [hunk 1]
```diff
@@ -130,8 +150,16 @@ class SequentialImpl : public Cloneable<SequentialImpl> {
...
-    for (const auto& module : modules_) {
-      clone->push_back(module.clone(device));
+    // Get the _registered_ names of submodules of this `Sequential`
+    //
+    // We don't want to use `modules_[i].ptr->name()` since this can return
+    // just the type name via RTTI. Since an `OrderedDict` cannot have duplicate
+    // key values, this will result in a runtime error if we have two submodules
+    // of the same type, e.g.:
+    // `Sequential sequential(Linear(3, 4), ReLU(), Linear(4, 1), ReLU())`
+    std::vector<std::string> registered_names = children_.keys();
```
[the remark this answers is not in the document]
Comment (kvathupo) [reply torch/csrc/api/include/torch/nn/modules/container/sequential.h:160]: Would prefer `const std::vector<std::string>` for clarity: saves descending the chain of Sequential->Cloneable->Module->OrderedDict to figure out the type

**Comment by kvathupo** [comment]
@malfet nudge- job failures seemingly unrelated, from jenkins

**Comment by kvathupo** [comment]
@malfet - ping

**Pull request closed without merging**